fe/src/components/Statistics/Statistics.re

```diff
@@ -0,0 +1,180 @@
+open Svc;
```

Comment:
Ogólnie do tego komponentu, to zauważyłem że ma on dwa problemy które mają chyba podobne źródło:
1. Problem. Lista pobranych userów oraz dane wykresu są trzymane w state w dwóch miejscach co stanowi niepotrzebną redundancję. Pierwsze miejsce to users: list(user) i chartData a drugie to LOADED(list(user), ...array(Js.Json.t))
2. Problem. Gdy klikamy w jakich checkbox z userem cała strona pokazuje kulki ładowania, podczas gdy pobierane są tylko dane do wykresu. Powoduje to takie nienaturalne znikanie na chwilę całej strony po kliknięciu w jakiegoś typka.

Według mnie rozwiązaniem tych dwóch jest takie:
1. Trzeba rozdzielić w stanie to czy są załadowaniu userzy i to czy są załadowane dane do serwisu.
2. Wtedy stan mógłby mieć taki kształt:

```
type usersState = 
| LOADING
| FAILURE
| LOADED(list(users));

type chartDataState = 
| LOADING
| FAILURE
| LOADED(array(Js.Json.t));

type state = {
  usersState,
  chartDataState,
  checkedUsers: list(int),
};
```

3. Teraz dzięki temu mógłyś zupełnie rodzielić renderowanie checkoxów od wykresu. render mógłby wyglądać jakoś tak:
```
{switch (state.chartDataState) {
  | FAILURE => <FailureMask />
  | LOADING => <LoadingMask />
  | LOADED(users) => wykres
  }}
 {switch (state.usersState) {
  | FAILURE => <FailureMask />
  | LOADING => <LoadingMask />
  | LOADED(users) => malujemy checboxy
  }}
```
Dzięki temu gdy klikalibyśmy w jakiś checbox tylko stan pobierania wykresu by się zmieniał a checboxy by pozostawały wyrenderowane. I pozbyłbyś się redundancji.

Reply:
To jest raczej spora architektnoniczna zmian. Jak chcesz można to zrobić oddzielnym PRem

Reply:
Myślę, że jak już zrobić, to porządnie. Albo chociaż w 3/4 porządnie :P. Ogarnę. Ale zapewne dopiero w sobotę albo niedzielę. Może jutro mi się uda.

Reply:
@kamilkuz 
Pozmieniałem nieco kod, ale mam pewien problem. Komponent `StatisticsChart` powinien wykonywać `send(LoadData)` za każdym razem, gdy otrzyma w propsach nową wersję `checkedUsersNids`. Nie wiem jak to zrobić. Próbowałem uzyskać jakiś sensowny wynik na podstawie [TEGO](https://github.com/reasonml/reason-react/blob/master/docs/lifecycles.md), ale coś mi nie szło :/.

Reply:
@ktrzos 
Masz problem bo nie zrobiłeś tak jak napisałem :P
Mi chodziło o to aby stan był tylko w jednym komponencie, np. w tym: StatisticsChartContainer, stan zawierający userów i dane statystyk, w formie w jakiej przedstawiłem, czyli:
```
type usersState = 
| LOADING_USERS
| FAILURE_USERS
| LOADED_USERS(list(users));

type chartDataState = 
| LOADING_CHART_DATA
| FAILURE_CHART_DATA
| LOADED_CHART_DATA(array(Js.Json.t));

// TO JEST TEN JEDEN STAN ZAWIERAJĄCY DANE USERÓW I DANE WYKRESU
type state = {
  usersState,
  chartDataState,
  checkedUsers: list(int),
};
```
Następnie elementy tego stanu byłyby przekazywane niżej przez propsy. Wtedy problem o którym napisałeś nie istnieje bo poprostu po kliknięciu w checkboxa wysyłasz akcję która wykona pobieranie danych, dane się pobiorą i same się przekażą niżej przez propsy i problem z głowy. To są właśnie zalety tego gdy jednak stan dla jakiejś funkcjonalności jest w jednym miejscu a nie zbyt bardzo rozproszony na komponentach. (Tak na marginesie właśnie tą zaletę wykorzystuje redux i szmagajofrejwork idąc krok dalej i robiąc jeden stan na całą aplikację).
Tutaj jeszcze taka filozoficzna rozkimna, według mnie stan skupiony w jednym miejscu ma swoje zalety i wady ale i stan rozproszony na zbyt wiele komponentów ma swoje zalety i wady, dlatego ważne jest według mnie zachowanie jakiejś równowagi w tym wszystkim - co często nie jest łatwe.

I jeszcze na marginesie nadal masz redundancje, bo dane userów trzymasz w LOADED(users) i w users.

Reply:
Co do tej redundancji, to fakt. Zapomniałem usunąć części kodu. Już jest jak być powinno.

Co do reszty. Jasne, zrozumiałem. Mimo wszystko starałem się właśnie rozproszyć jakoś ten stan. Tak jak dane `users` wyciągnąłem wyżej, bo są one potrzebne w dwóch podrzędnych komponentach, tak dane do rysowania wykresu potrzebne są tylko w jednym z nich. Stąd zdecydowałem o przeniesieniu logiki wysyłania serwisu z zapytaniem o dane właśnie tam. Czy to właśnie nie jest dobra praktyka, skoro te informacje potrzebne są tylko i wyłącznie w tym miejscu? No i też jest chyba dobrą praktyką, aby coś się działo po wpłynięciu nowej wersji propsów, bo to jednak wspierane przez Reacta :).

Plusem tego jest również odchudzenie komponentu `StatisticsChartContainer`. Ma mniej kodu, a i jest więc bardziej czytelny. Dla mnie im mniejsza ilość kodu w jednym pliku, tym lepiej.

Reply:
W sumie to po części zrobiłem jak napisałeś, ale te stany odosobniłem do poszczególnych komponentów :).

Reply:
Dodałem do komponentu `StatisticsChart` "willReceiveProps" i to działa. Co o tym myślisz?

Reply:
@ktrzos 
No właśnie, niby rozdrabnianie jest dobre a tu jeb jest teraz problem - jak wymusić zmianę stanu w komponencie który jest dzieckiem. W reackie przepływ jest jednostronny, propsy przepływają z góry na dół, a akcje można wyoływać z dołu do góry (nigdy z góry do dołu) - można robić oczywiście jakieś 'sztuczki' to co próbowałeś czyli po zmianie propsów wywoływać jakieś akcje - ale dla mnie to są takie lekkie hacki. I waśnie często się robi tak, że dla jednej funkcjonalności która składa się z wielu komponentów stan umieszcza się tylko w jedym miejscu, najbardziej parentowym dla tej funkjonalności. Według mnie to jest poprawna architektura.
To jest właśnie to o czym mówiłem że właśnie to wykorzystuje redux w swoim podejściu robiąc jeden parentowy stan.
Jeśli przeszkadza Ci to że plik który zawiera ten stan jest duży to nie widzę problemu aby wyciągać jakieś funkcje (grupy funkcj) do osobnych plików.

Reply:
@ktrzos 
I jeszcze takie dopowiedzenie odnośnie tego że dane potrzebne do rysowania wykresu są potrzebne tylko w komponencie z wykresem. Owszem potrzebne są tylko w komponencie który rysuje wykres, ale akcja powodująca pobieranie tych danych już pochodzi spoza tego komponentu (kliknięcie w checkbox jest poza tym komponentem) i właśnie to sprawia że stan tego wykresu powinien być w parencie a nie w nim.

Reply:
> ale akcja powodująca pobieranie tych danych już pochodzi spoza tego komponentu (kliknięcie w checkbox jest poza tym komponentem) i właśnie to sprawia że stan tego wykresu powinien być w parencie a nie w nim.

To w sumie mnie mega przekonało. Masz rację. I tak i tak jest to od siebie zależne przez te checkboxy. No widzisz, muszę się nauczyć takiego podejścia.

> Jeśli przeszkadza Ci to że plik który zawiera ten stan jest duży to nie widzę problemu aby wyciągać jakieś funkcje (grupy funkcj) do osobnych plików.

Prawda, zawsze tak przecież można zrobić. No ok, spróbuję to ogarnąć :).

Reply:
Spoczko :)
Jeszcze tylko odniosę się do tego co napisałeś, w celach edukacyjnych:

> No i też jest chyba dobrą praktyką, aby coś się działo po wpłynięciu nowej wersji propsów, bo to jednak wspierane przez Reacta :).
ReasonReactowy willReceiveProps jest tylko bindingiem do reactowego: componentWillReceiveProps który jeszcze jest wspierany w reakcie 16 ale w 17 już nie będzie - nawet zmienili jego nazwę na: UNSAFE_componentWillReceiveProps()
Wszystko jest tu: https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops

To tylko tak edukacyjnie :)

Reply:
@kamilkuz 
Ok, FE zrobione... mam nadzieję. Spójrz proszę na to. Liczę, że chociaż w większej części będzie to już odpowiednie. Ewentualne, mniejsze poprawki przewidywalne.

Reply:
No i teraz jest elegancko :)
Nawet mam jeszcze pomysł jak można zrobić aby po kliknięciu nie znikał całkowicie na chwilkę wykres, ale tak jak jest teraz jest całkowicie akceptowalne.
A nawet mam też pomysł jak byś bardzo chciał w jaki sposób rodzielić te stany na dwa niezależne (teraz dopiero wpadłem na to, da się bez willReceiveProps tylko trzeba odpowiednio podzielić ten stan, na dwa, i tu kuczowe: zupełnie niezależne stany) - ale to jak będziesz chcial to jutro na żywo mogę wytłumaczyć, bo znowu wyjdzie dużo pisania. 

No ale tak jak pisałem teraz jest jak najbardziej spoko :)

Reply:
@kamilkuz 
Zrobiłem aby wykres nie znikał przy zmianie userów :).
Co do innego sposobu wykonania, to możemy pogadać np. w czwartek, jak ogarnę tą auto-publikację w Leonie. Ale to ewentualnie osobnym PR-em by się zrobiło, jeśli trzeba będzie.

Reply:
@ktrzos 
Spoko ten loading, właśnie coś takiego chciałem zaproponować.
A co do innego sposobu podziału stanu to jak najbardziej można na żywo pogadać